### DISCUSSION TITLE
Tinker9: Error in make step

### DISCUSSION THREAD

**Discussion by larsakth** [TinkerTools/tinker-gpu#167] [Q&A] [answered]
Hi,
I am trying to compile the gpu-version of tinker9. CentOS 7.8 as platform, and using NVIDIA V100 cards.

I get an error in the make step (make -j). Cmake (v. 3.18.1) builds the makefile without errors. I use nvfortran and nvc (v. 21.7). The final part of the output with the error message looks like this:

...
[ 92%] Built target tinker9_libtinker
make[1]: *** [CMakeFiles/__t9_all_tests_o.dir/all] Error 2
make: *** [all] Error 2

Any idea on how to solve this?

Thanks!
Lars

**Comment by zhi-wang** [comment]
Hi, if you just run "make" (without -j), it will give you a more readable error message. Without it, I can only guess what happened.

My guess is that I don't think nvfortran will work with Tinker9 due to a long-time bug in the compiler. I will add more information in the FAQ about this problem.

gfortran or ifort will work properly with nvcc and nvc++ if they have been configured correctly. The c-compiler, nvc, is not used here.

**Comment by larsakth** [comment, marked as answer]
Thanks Zhi,

I tried your suggestions with:
- export FC=gfortran CC=nvcc
- cmake ...
   (wth the recommended flags and directories; a few warnings but no errors)
   (however, if instead exporting CC=nvc++, cmake halts complaining that nvc++ cannot compile the test file)
   A Makefile is produced, as expected
- make
   (without -j)
   runs well until the very end giving the following error message:
...
[100%] Built target tinker9_cpp
Scanning dependencies of target tinker9
/opt/tinker9/build-cmake/src/libtinker9_cpp.a(info_x.cc.o): In function `tinker::x_info(int, char**)':
/opt/tinker9/src/info_x.cc:88: undefined reference to `tinker::cuda_compiler_name()'
pgacclnk: child process exit status 1: /usr/bin/ld
make[2]: *** [CMakeFiles/tinker9] Error 2
make[1]: *** [CMakeFiles/tinker9.dir/all] Error 2
make: *** [all] Error 2

Any idea what goes wrong?

Best regards,
Lars

**Reply by zhi-wang** [reply]
Hi, the function `tinker::cuda_compiler_name()` is implemented in `src-cu/compilers_cu.cu`. I'm not sure if this source code gets compiled.

I assume you are using `gfortran` for the `.f` files, `nvc++` for the `.cpp` files under `src` and `src-acc`, and `nvcc` for the `.cu` files.  (For the `.cpp` files under `src`, you can use `g++` too. But this requires extra configuration in the setup and I assume this is not what you are doing.) `CC` stands for "C Compiler" in CMake and Makefile. Tinker9 does not use any C compiler, so the `CC` flag should not be set at all.

**Reply by larsakth** [reply]
Hi, yes, I use the combination of gfortran, nvc++ and nvcc; the latter two of the nvhpc version 21.5 bundled with cuda 11.3 (the cuda version installed on the machine). Thus, it should work. In fact, 'compilers_cu.cu' gets compiled to an object file. It is at the very final linking stage that something goes wrong.

I have tried to set various environmental variables and flags to point at suitables libraries for the linking. However, this has not resulted in any better result. Any ideas on how to move forward?

**Reply by zhi-wang** [reply]
Hi, the source files have been renamed and the CMake files were rewritten a few weeks ago. If you are still having `*_cu.cu` files, you should consider updating the source code and then see what happens next. Not sure if this would fix your problem though.

**Reply by larsakth** [reply]
Hi, I downloaded the new Dec 6 version and made a compilation attempt with a similar result (see below) --- different program but same error message. The error output from 'make' is given below:

[ 94%] Building CXX object CMakeFiles/tinker9.dir/src/main_tinker9.cc.o
/opt/tinker9/src/main_tinker9.cc:
[ 95%] Linking CXX executable tinker9
src/CMakeFiles/tinker9_version.dir/info_x.cc.o: In function `tinker::x_info(int, char**)':
/opt/tinker9/src/info_x.cc:92: undefined reference to `tinker::cuda_compiler_name()'
pgacclnk: child process exit status 1: /usr/bin/ld
make[2]: *** [tinker9] Error 2
make[1]: *** [CMakeFiles/tinker9.dir/all] Error 2
make: *** [all] Error 2

I recently solved a similar linker-related problem when compiling Quantum Espresso for GPU by adding a linker flag in the Makefile, but there I got some hints to the problems from the compilation errors. The problems here are less clear to me.

**Reply by zhi-wang** [reply]
I still have no idea what happened here. It looks like a linker problem to me. Would you paste your results for the following procedures?

- What is your current `g++` version?
- My build directory is `~/tinker9/build`. Can you find the following object files in similar locations? My files are at:
    - [ ] `~/tinker9/build/src/CMakeFiles/tinker9_version.dir/info_x.cc.o`
    - [ ] `~/tinker9/build/src/cu/CMakeFiles/tinker9_cu.dir/compilers.cu.o`
- What are the results for the `nm -n path_to/info_x.cc.o path_to/compilers.cu.o` command? Do we have matching results here?
    - [ ] `info_x.cc.o`
    - [ ] `compilers.cu.o`
```
nm -n info_x.cc.o
...
                 U _ZN6tinker17acc_compiler_nameB5cxx11Ev
                 U _ZN6tinker17cxx_compiler_nameB5cxx11Ev
                 U _ZN6tinker18cuda_compiler_nameB5cxx11Ev
...

nm -n compilers.cu.o
...
0000000000000000 b _ZL20__cudaFatCubinHandle
0000000000000000 t _ZL24__sti____cudaRegisterAllv
0000000000000000 t _ZL26__cudaUnregisterBinaryUtilv
0000000000000010 T _ZN6tinker18cuda_compiler_nameB5cxx11Ev
...
```

**Reply by larsakth** [reply]
Dear Zhi,

Short answers/output below:

- What is your current `g++` version?
(The standard g++ is 9.3.0)
The used nvc++ version is 21.5 (bundled with cuda version 11.3; installed)

- My build directory is `~/tinker9/build`. Can you find the following object files in similar locations? My files are at:
    - [ ] `~/tinker9/build/src/CMakeFiles/tinker9_version.dir/info_x.cc.o`
    - [ ] `~/tinker9/build/src/cu/CMakeFiles/tinker9_cu.dir/compilers.cu.o`
Essentially same as yours:
   /opt/tinker9/build-cmake/src/CMakeFiles/tinker9_version.dir/info_x.cc.o
   /opt/tinker9/build-cmake/src/cu/CMakeFiles/tinker9_cu.dir/compilers.cu.o

- What are the results for the `nm -n path_to/info_x.cc.o path_to/compilers.cu.o` command? Do we have matching results here?
    - [ ] `info_x.cc.o`
    - [ ] `compilers.cu.o`
```
No … seems like information is missing for this one:
nm –n info_x.cc.o
…
                 U _ZN6tinker17acc_compiler_nameEv
                 U _ZN6tinker17cxx_compiler_nameEv
                 U _ZN6tinker18cuda_compiler_nameEv
…
Yours:
nm -n info_x.cc.o
...
                 U _ZN6tinker17acc_compiler_nameB5cxx11Ev
                 U _ZN6tinker17cxx_compiler_nameB5cxx11Ev
                 U _ZN6tinker18cuda_compiler_nameB5cxx11Ev
...
But, yes, seems ok for this one:
nm –n compilers.cu.o
…
0000000000000000 b _ZL20__cudaFatCubinHandle
0000000000000000 t _ZL24__sti____cudaRegisterAllv
0000000000000000 t _ZL26__cudaUnregisterBinaryUtilv
0000000000000010 T _ZN6tinker18cuda_compiler_nameB5cxx11Ev
…
Yours:
nm -n compilers.cu.o
...
0000000000000000 b _ZL20__cudaFatCubinHandle
0000000000000000 t _ZL24__sti____cudaRegisterAllv
0000000000000000 t _ZL26__cudaUnregisterBinaryUtilv
0000000000000010 T _ZN6tinker18cuda_compiler_nameB5cxx11Ev
...
```

**Reply by zhi-wang** [reply]
It seems that `nvc++` didn't use the cxx11 ABI for some reason. I have never added/changed any related flags in the building scripts. I'd do the following checks.

Create a file in a directory, then compile it with `g++`, `nvc++`, and `nvcc`.
```
// test.cpp
#include <string>
namespace tinker {
std::string cuda_compiler_name() { return ""; }
}

g++ z.cpp -std=c++11 -O3 -c -o g.o
nvc++ z.cpp -std=c++11 -O3 -c -o nv.o
nvcc z.cpp -std=c++11 -O3 -c -o cu.o
nm -n g.o nv.o cu.o

# my result (as expected)
g.o:
0000000000000000 T _ZN6tinker18cuda_compiler_nameB5cxx11Ev
nv.o:
0000000000000000 T _ZN6tinker18cuda_compiler_nameB5cxx11Ev
cu.o:
0000000000000000 T _ZN6tinker18cuda_compiler_nameB5cxx11Ev
```

This will check if the problem is in `nvc++` itself or is inherited from `g++`. I'd assume your object files made by the `g++` and `nvcc` are good but `nvc++` will be different. I noticed that the cxx11 ABI is not used by default until g++ 5. I would guess that when `nvc++` was installed on the machine, the default g++ was still version 4 (likely 4.8.5). The configuration file can be cat-ed:
`cat $(dirname $(which nvc++))/localrc`. It will help if you can paste the result here too. My file is

```
cat $(dirname $(which nvc++))/localrc

set LFC=-lgfortran;
set LDSO=/lib64/ld-linux-x86-64.so.2;
set GCCDIR=/usr/lib/gcc/x86_64-linux-gnu/9;
set G77DIR=/usr/lib/gcc/x86_64-linux-gnu/9/;
set OEM_INFO=64-bit target on x86-64 Linux $INFOTPVAL;
set GNUATOMIC=-latomic;
set GCCINC= /usr/lib/gcc/x86_64-linux-gnu/9/include /usr/local/include /usr/include/x86_64-linux-gnu /usr/include;
set GPPDIR= /usr/include/c++/9 /usr/include/x86_64-linux-gnu/c++/9 /usr/include/c++/9/backward /usr/lib/gcc/x86_64-linux-gnu/9/include /usr/local/include /usr/include/x86_64-linux-gnu /usr/include;
set LOCALRC=YES;
set EXTENSION=__extension__=;
set COMPGCCINCDIR=include-gcc70;
set LC=$if(-Bstatic,-lgcc -lgcc_eh -lc -lgcc -lgcc_eh -lc, -lgcc -lc -lgcc -lgcc_s);
set DEFSTDOBJDIR=/usr/lib/x86_64-linux-gnu;
# GLIBC version 2.31
# GCC version 9.3.0
set GCCVERSION=90300;
export PGI=$COMPBASE;
# makelocalrc executed by root Fri 08 Jan 2021
```

If this indeed shows an old version on the line `set GCCVERSION`, these links will be helpful:

https://github.com/TinkerTools/tinker9/discussions/140#discussioncomment-568841
https://github.com/TinkerTools/tinker9/blob/master/doc/manual/m/install/preq.rst

**Reply by larsakth** [reply]
Thanks!

It seems to be a problem with the nvc++. My output from 'nm -n g.o nv.o cu.o' is:

g.o:
0000000000000000 T _ZN6tinker18cuda_compiler_nameB5cxx11Ev

nv.o:
                 U _ZNSsC1EPKcRKSaIcE
0000000000000000 r .S03775
0000000000000000 T _ZN6tinker18cuda_compiler_nameEv

cu.o:
0000000000000000 T _ZN6tinker18cuda_compiler_nameB5cxx11Ev

And it looks that you are quite right --- the output of 'cat ...' being:
# GCC version 4.8.5
set GCCVERSION=40805;

I will look into what can be done there.

Best regards,
Lars

**Reply by zhi-wang** [reply]
Is this problem fixed?

**Reply by larsakth** [reply]
Hello Zhi, thanks for asking. I edited the localrc file to point at the right directories wrt a newer version of gcc (v.9.3.0). That also involved installation of a newer glibc (v.2.32). After those actions, tinker9 compiled without error messages. I have asked a co-worker to test the program. I will be back once I get some feedback. Best regards, Lars

**Reply by larsakth** [reply]
Hello Zhi,
We have now tried to run the program. However, we face problems that involve 'Permission denied' (even as root). I cannot find any reasons relating to the shell nor to how the file system has been mounted. Any ideas?
Best regards, Lars

**Reply by zhi-wang** [reply]
This is also new to me. How did you run the executable? Are you able to locate the tinker9 and all.tests executables? Can you exec them directly in your build directory?

**Reply by larsakth** [reply]
Yes, both executables (tinker9 and all.tests) are formed in the compilation without any erors, and they can be found in the build directory (with correct permissions etc.) together with a few scripts for the different types of runs. Output below (make.log is the output from the compilation; calc is a group with all users allowing access to running all programs in /opt):

-rwxr-xr-x.  1 root calc 43692696 Dec 26 09:35 all.tests
-rwxr-xr-x.  1 root calc       78 Dec 26 09:30 analyze9.sh
-rwxr-xr-x.  1 root calc       74 Dec 26 09:30 bar9.sh
-rw-r--r--.  1 root calc    15876 Dec 26 09:19 CMakeCache.txt
drwxr-xr-x. 13 root calc     4096 Dec 26 09:35 CMakeFiles
-rw-r--r--.  1 root calc     2106 Dec 26 09:19 cmake_install.cmake
-rwxr-xr-x.  1 root calc       78 Dec 26 09:30 dynamic9.sh
drwxr-xr-x.  3 root calc     4096 Dec 26 09:19 ext
-rwxr--r--.  1 root calc       40 Dec 26 09:53 hello.test
-rwxr-xr-x.  1 root calc       75 Dec 26 09:30 info9.sh
-rw-r--r--.  1 root calc    13300 Dec 26 09:19 Makefile
-rw-r--r--.  1 root calc   420774 Dec 26 09:35 make.log
-rwxr-xr-x.  1 root calc       79 Dec 26 09:30 minimize9.sh
drwxr-xr-x.  7 root calc     4096 Dec 26 09:19 src
-rwxr-xr-x.  1 root calc       79 Dec 26 09:30 testgrad9.sh
-rwxr-xr-x.  1 root calc 39555312 Dec 26 09:33 tinker9

Execution in the build directory also gives the 'Permission denied' message.

**Reply by zhi-wang** [reply]
I have no idea what could cause that. Is it correct that permission is denied for both of the real executables and the shell scripts? Can you compile a "hello world" with C++ (nvc++/nvcc/g++) and run them in this build directory? This is weird.

**Reply by larsakth** [reply]
Thanks, a good idea. And yes, the results are a bit interesting: nvcc and g++ produce an executable that works properly. However, nvc++ does not --- Compilation works without errors and produces an executable intermediate in size wrt g++ and nvcc, but when running the compiled  program 'Permission denied' emerges just as for the tinker9 executables. I will test a few different options regarding the c++ compilation of tinker9.

**Reply by zhi-wang** [reply]
I am really out of ideas. This is definitely related to the installation of the NVHPC toolchain. I encountered and solved many problems with it, and I can understand things are more complex and out of our control on the big clusters. One place I found very helpful is https://forums.developer.nvidia.com/c/accelerated-computing/hpc-compilers/nvc-nvc-and-nvfortran/313 -- It used to be a forum on the PGI website and now it is migrated to Nvidia. I'd post this "hello world" problem there and ask for ideas.

**Reply by larsakth** [reply]
I could trace the problem to an installation of a newer version of glibc than that coming with the Centos 7 system. Most likely, this caused some confusion between the two versions upon compiling/running the program. Backing off from the new glibc version actually allowed compilation and successful test runs on both A100 and V100 cards. The compiler combination eventually working was gcc/g++ 9.3.0, nvcc of Cuda 11.3 and nvc++ from Nvidia hpc_sdk 21.5 (bundled with Cuda 11.3, and where the localrc file needs to be modified).
Many thanks for all your help!
Lars
